feat(docs): update TEE extension guide (#1424)
flare-foundation/developer-hub: docs/fcc/guides/01-sign.mdx

Build and deploy a **Trusted Execution Environment (TEE) extension** that securely stores a private key and signs arbitrary messages.
This guide walks through every step — from writing the smart contract and extension handler to deploying on Coston2 and running an end-to-end test.
The code for this example is available on [GitHub](https://github.com/flare-foundation/fce-sign).

:::warning[Demonstration only]
This example is for learning purposes.
Storing encrypted secrets on-chain is not advisable in production — on-chain data is public and encryption can be broken over time.
A production extension should use offchain channels for secret delivery.
:::

:::info[New to Flare TEE?]
A TEE extension is an offchain program that runs inside a Trusted Execution Environment.
It receives **instructions** from on-chain transactions, processes them in a secure enclave, and writes the results back on-chain.
The TEE framework handles attestation, key management, and message routing — you only write the business logic.
:::

## Overview

The Private Key Manager extension demonstrates the core TEE workflow:

> 1. A user sends an [Elliptic Curve Integrated Encryption Scheme (ECIES)](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme) encrypted private key onchain via the `InstructionSender` contract.

Reviewer: ```suggestion
1. A user sends an [Elliptic Curve Integrated Encryption Scheme (ECIES)](https://en.wikipedia.org/wiki/Integrated_Encryption_Scheme) encrypted private key on-chain via the `InstructionSender` contract.
```

2. The TEE extension decrypts and stores the key inside the secure enclave.
3. A user sends a `sign` instruction with an arbitrary message.
4. The TEE extension signs the message with the stored key and returns the signature on-chain.

We will build this in three parts:

> 1. **onchain contract** that sends instructions;

Reviewer: ```suggestion
1. **on-chain contract** that sends instructions;
```

> 2. **offchain handler** that processes them;

Reviewer: ```suggestion
2. **off-chain handler** that processes them;
```

- **`extension-tee`:** Your extension code (Go, Python, or TypeScript).
  Receives decoded instructions from the proxy and returns results.
- **`ext-proxy`:** The TEE extension proxy.
  Watches the chain for new instructions targeting your extension, forwards them to your handler, and submits results back on-chain.
- **`redis`:** In-memory store used by the proxy for internal state.

## Prerequisites

Before you begin, make sure you have the following installed:

- [Docker Desktop](https://docs.docker.com/get-docker/).
- [Foundry](https://book.getfoundry.sh/getting-started/installation) for contract compilation.
- [Go](https://go.dev/dl/) — for deploy and registration tools in `go/tools/`.

> - [ngrok](https://ngrok.com/) — to expose your local proxy to the internet.

Reviewer: Is this a hard dependency, or can you still choose between it and Cloudflare?

Author: Right now in repo guide only ngrok is used

- A funded Coston2 wallet with C2FLR for gas and TEE registration fees — use the [Coston2 faucet](https://faucet.flare.network/coston2).

## Onchain Contract

The `InstructionSender` contract is the onchain entry point.
It interacts with two Flare system contracts:

- **`TeeExtensionRegistry`:** Registers extensions and routes instructions to TEE machines.
- **`TeeMachineRegistry`:** Tracks registered TEE machines and provides random selection.

### Contract Code

<CodeBlock language="solidity" title="contracts/InstructionSender.sol">
  {InstructionSender}
</CodeBlock>

:::note
The constructor takes the addresses of the two Flare system contracts.

> These are already deployed on Coston2 — the deploy tooling reads their addresses from `config/coston2/deployed-addresses.json`.

Reviewer: ```suggestion
These are already deployed on Coston2 — the deployment tooling reads their addresses from `config/coston2/deployed-addresses.json`.
```

 }

    encoded, err := abiEncodeTwo(df.OriginalMessage, sig)
    if err != nil {
        return buildResult(action, df, nil, 0, fmt.Errorf("ABI encoding failed: %v", err))
 }

    return buildResult(action, df, encoded, 1, nil)
}
```

The decryption uses the TEE node's built-in `/decrypt` endpoint on `SIGN_PORT` (default 7701).
The caller ECIES-encrypts the private key using the TEE's public key (fetched from the proxy's `/info` endpoint) before sending it on-chain.

### Python and TypeScript implementations

Python and TypeScript use a `Framework` that registers handlers by `opType`/`opCommand` pair:

```python title="python/app/handlers.py"
def register(framework: Framework) -> None:
 framework.handle(OP_TYPE_KEY, OP_COMMAND_UPDATE, handle_key_update)
 framework.handle(OP_TYPE_KEY, OP_COMMAND_SIGN, handle_key_sign)
```

Each handler receives a hex-encoded `originalMessage` and returns `(data, status, error)`:

- **`data`:** Hex-encoded return data written back on-chain, or `None` if no data;
- **`status`:** `0` = error, `1` = success, `>=2` = pending;
- **`error`:** Error message string if the handler failed.

> The `handle_key_update` handler decrypts via the TEE node and stores the key; `handle_key_sign` signs the message and ABI-encodes the result — the same logic as the Go implementation, adapted to the Framework API.

Reviewer: I feel like there's a word missing after the `The `handle_key_update` handler decrypts`

├── scripts/                     # Primary deploy interface
│   ├── use-chain.sh
│   ├── pre-build.sh
│   ├── start-services.sh
│   ├── post-build.sh
│   └── test.sh
├── go/
│   ├── internal/extension/      # Go business logic (modify these)
│   └── tools/cmd/               # All deploy CLIs (Go-only)
├── python/                      # Python implementation
│   ├── app/                     # Your business logic (modify these)
│   └── base/                    # Framework infrastructure (do not modify)
├── typescript/                  # TypeScript implementation
│   ├── src/app/                 # Your business logic (modify these)
│   └── src/base/                # Framework infrastructure (do not modify)
├── proxy/                       # Extension proxy (Docker build context)
├── docker-compose.yaml
├── .env.example
├── .env.local.coston2           # Local simulated Coston2 template
└── README.md
```

## Deploying and Testing on Coston2

This walkthrough deploys a **local simulated TEE** against the real Coston2 chain using Docker and an ngrok tunnel.
For production deployment on a GCP Confidential Space VM, see `DEPLOYMENT_STEPS.md` in the repository.

:::tip[Quick start]

> After configuring keys, ngrok, and the indexer TOML (steps 0-4 below):

Reviewer: ```suggestion
After configuring keys, ngrok, and the indexer TOML (steps 0-4 below), run the following scripts:
```

Or whatever is appropriate here.

```bash

> ./scripts/pre-build.sh

Reviewer: ```suggestion
- `./scripts/pre-build.sh`
```

And I'd do the same for other three.

Select the local simulated Coston2 environment and extension language:

```bash
./scripts/use-chain.sh local coston2 go
```

This copies `.env.local.coston2` to `.env`, setting `SIMULATED_TEE=true` and `LOCAL_MODE=false` (you are on the real Coston2 chain; only the TEE attestation is simulated).

Use `python` or `typescript` instead of `go` to run a different extension image.
Run `./scripts/use-chain.sh --list` to see available chains and languages.

### Step 1: Configure deployer keys

Edit `.env.local.coston2` and set your funded Coston2 credentials:

```bash title=".env.local.coston2"
DEPLOYMENT_PRIVATE_KEY="<your-funded-coston2-private-key-hex-no-0x>"
INITIAL_OWNER="0x<your-address>"
```

Re-activate after editing so `.env` picks up the changes:

```bash
./scripts/use-chain.sh local coston2 go
```

### Step 2: Deploy contract and register extension

```bash
./scripts/pre-build.sh
```

This compiles Solidity, deploys `InstructionSender`, and registers the extension on-chain.
On success, it writes `EXTENSION_ID` and `INSTRUCTION_SENDER` to `config/extension.env` — do not edit this file manually unless recovering from an error.

:::warning

> `pre-build.sh` refuses to re-run if `config/extension.env` already exists.

Reviewer: ```suggestion
The `pre-build.sh` script refuses to re-run if `config/extension.env` already exists.
```

This is just a personal preference of mine, coming from a course on writing mathematical articles; where they told us not to start sentences with plain math-style symbols. I consider the code snippets as such, so I prefer no to start sentences with them.

> Use `./scripts/pre-build.sh --force` only when you intentionally need a new extension.

Reviewer: Either

```suggestion
Use `./scripts/pre-build.sh --force` only when you intentionally want a new extension.
```

or

```suggestion
Use `./scripts/pre-build.sh --force` only when you genuinely  need a new extension.
```

> Re-running against an existing TEE orphans it and causes `test.sh` to fail with `MachineManager.TooMany()`.

Reviewer: I don't understand what `orphans` means in this context.

:::

### Step 3: Start the ngrok tunnel

In a separate terminal, expose the proxy's public port:

```bash
ngrok http 6674
```

Copy the HTTPS URL from ngrok's **Forwarding** line and set it in `.env.local.coston2`:

```bash title=".env.local.coston2"
EXT_PROXY_URL="https://<your-ngrok-domain>"
```

Re-activate the environment:

```bash
./scripts/use-chain.sh local coston2 go
```

:::info

> ngrok's free tier provides one reserved domain, so this URL normally stays stable across restarts.

Reviewer: ```suggestion
The `ngrok` free tier provides one reserved domain, so this URL normally stays stable across restarts.
```

Only update `EXT_PROXY_URL` if the domain changes.
:::

### Step 4: Configure the indexer database

The local `ext-proxy` queries Flare's C-chain indexer to find TEE events.
Create the Coston2 proxy config from the bundled example:

```bash
cp config/proxy/extension_proxy.coston2.docker.toml.example \
 config/proxy/extension_proxy.coston2.docker.toml
```

Edit the `[db]` block in `config/proxy/extension_proxy.coston2.docker.toml` with the Coston2 indexer host and the read-only credentials provided to you (not published in this guide):

```toml title="config/proxy/extension_proxy.coston2.docker.toml"
[db]
host = "34.38.42.208"
port = 3306
database = "indexer"
username = "<your-indexer-db-username>"
password = "<your-indexer-db-password>"
log_queries = false
```

> The Coston2 indexer is reachable without VPN.

Reviewer: ```suggestion
The Coston2 indexer is reachable without a VPN.
```

> Coston uses different credentials and requires Abelium VPN — see `DEPLOYMENT_STEPS.md` in the repository if deploying to Coston instead.

Reviewer: Maybe avoid mentioning `Abelium`? I'm imagining this is mentioned as a comment in code and Claude picked it up.

:::info[Flare Indexer Access]

> To get the indexer credentials, please get in touch with us via [support](https://flare.network/support) or [X](https://x.com/FlareDevs) and share what you are building.

Reviewer: ```suggestion
To get the indexer credentials, please get in touch with us via [support](https://flare.network/support) or [X](https://x.com/FlareDevs), and share what you are building.
```

:::

### Step 5: Start the extension stack

```bash
./scripts/start-services.sh
```

This builds the extension image for your `LANGUAGE`, then starts redis, ext-proxy, and extension-tee.
Wait for the proxy to become healthy:

```bash
until curl -sf http://localhost:6674/info >/dev/null 2>&1; do sleep 2; done
echo "Extension proxy is ready"
```

:::tip[Switching languages]
Change `LANGUAGE` in `.env.local.coston2`, re-run `use-chain.sh`, then re-run `start-services.sh` to rebuild the extension image.
:::

### Step 6: Verify the proxy

Confirm the simulated TEE is reporting the correct extension ID:

```bash
curl -s "$EXT_PROXY_URL/info" | jq '.machineData'
```

For a simulated TEE, expect:

| Field          | Expected                                         |
| -------------- | ------------------------------------------------ |
| `codeHash`     | Simulated hash (`0x194844cf…`)                   |
| `extensionId`  | Matches `EXTENSION_ID` in `config/extension.env` |
| `initialOwner` | Matches your `INITIAL_OWNER`                     |

### Step 7: Register the TEE machine

```bash
./scripts/post-build.sh
```

This runs two onchain steps:

> 1. **`allow-tee-version`** — allows the code hash and platform for your extension.

Reviewer: ```suggestion
1. **`allow-tee-version`:** allows the code hash and platform for your extension.
```

> 2. **`register-tee -command rRap`** — pre-registers the TEE, requests fresh attestation, runs the FTDC availability check, and promotes to production.

Reviewer: ```suggestion
2. **`register-tee -command rRap`:** pre-registers the TEE, requests fresh attestation, runs the FTDC availability check, and promotes to production.
```

The capital `R` in `rRap` issues a fresh attestation challenge on re-runs, avoiding `Verification.ChallengeExpired`.

> `SIMULATED_TEE=true` in your `.env` pairs with `MODE=1` injected by Docker Compose for simulated attestation.

Reviewer: ```suggestion
The `SIMULATED_TEE=true` variable in your `.env` pairs with `MODE=1` injected by Docker Compose for simulated attestation.
```

### Step 8: Run the end-to-end test

```bash
./scripts/test.sh
```

The test performs the following sequence:

1. Calls `setExtensionId()` on the `InstructionSender` contract to discover and store the extension ID.
2. Fetches the TEE's public key from the proxy's `/info` endpoint.
3. ECIES-encrypts a test private key using the TEE's public key.
4. Sends an `updateKey` instruction on-chain with the encrypted key.
5. Waits for the TEE to process the instruction and store the key.
6. Sends a `sign` instruction on-chain with a test message.
7. Verifies the returned ECDSA signature matches the test private key.

If the test passes, your extension is fully operational.

## Port Reference

| Service            | Container port | Host port |
| ------------------ | -------------- | --------- |
| ext-proxy internal | 6663           | 6673      |
| ext-proxy external | 6664           | 6674      |
| redis              | 6379           | 6382      |

> The ngrok tunnel exposes host port **6674** (ext-proxy external) to the internet.

Reviewer: ```suggestion
The `ngrok` tunnel exposes host port **6674** (ext-proxy external) to the internet.
```

Ensure your wallet has enough C2FLR for gas and fees.
Fund it with the [Coston2 faucet](https://faucet.flare.network/coston2).

**`MachineManager.TooMany()` during test**

The extension ID in `config/extension.env` does not match the on-chain TEE record.

**`Verification.ChallengeExpired`**

Re-run `post-build.sh`.

**`code hashes do not match`**

> `SIMULATED_TEE` and the container `MODE` env disagree.

Reviewer: ```suggestion
The `SIMULATED_TEE` variable and the container `MODE` env disagree.
```

Also, `env`?

```

If that doesn't help, the FDC attestation flow requires active relay providers on Coston2.

> **ngrok URL changed**

Reviewer: ```suggestion
**`ngrok` URL changed**
```

This is one place where I wouldn't try to fit in an extra in front of the code snippet.

1. Update `EXT_PROXY_URL` in `.env.local.coston2` and re-run `use-chain.sh`.
2. Restart the Docker stack: `./scripts/stop-services.sh && ./scripts/start-services.sh`.
3. Re-run `post-build.sh` and `test.sh`.

## Cleanup

**Stop the Docker stack**

```bash
./scripts/stop-services.sh
```

This stops and removes all containers (redis, ext-proxy, extension-tee).

**Full reset**

> To completely reset local state and start from scratch:

Reviewer: ```suggestion
To completely reset the local state and start from scratch:
```

```bash
./scripts/stop-services.sh
docker compose down --rmi local
rm -f .env config/extension.env config/proxy/extension_proxy.coston2.docker.toml
```

After a full reset, start again from [Step 0](#step-0-activate-local-simulated-mode).

To mint a new extension after a diamond redeploy without a full reset:

```bash
./scripts/pre-build.sh --force
```

:::info

> Onchain state (deployed contracts, registered extensions, registered TEEs) cannot be reset.

Reviewer: ```suggestion
On-chain state (deployed contracts, registered extensions, registered TEEs) cannot be reset.
```
